### PR TITLE
fix(unifiedsearch): open file requested via traymenu-searchbar directly and locally 

### PR DESCRIPTION
**Pull request by barisbasar1209**
This pull request addresses issue #7814. 

**Previously:** 

Whenever the user searches for and requests a file or folder through the traymenu-searchbar... 

- Nextcloud Server Version 20: ...the containing folder is opened in the local file explorer, instead of the requested file or folder itsself.
- Nextcloud Server Version above 20: ...the requested file or folder is opened, but in the webinterface filesystem application. 

**Change:** 

For any server version using the unified search (20+) the requested file or folder is opened in the local file explorer 

Authors are @mike0609king and me. 

Thanks a lot in advance for review and suggestions!

### CODE REVIEW

File: test/testunifiedsearchlistmodel.cpp [hunk 1]
```diff
@@ -575,13 +575,22 @@ private slots:
...
                     model->data(model->index(i), OCC::UnifiedSearchResultsListModel::DataRole::ProviderIdRole)
                         .toString();
+
+                const QString subline = 
+                    model->data(model->index(i), OCC::UnifiedSearchResultsListModel::DataRole::SublineRole)
+                        .toString();
+
+                const QString title = 
+                    model->data(model->index(i), OCC::UnifiedSearchResultsListModel::DataRole::TitleRole)
+                        .toString();
```
Comment (mgallien) [inline test/testunifiedsearchlistmodel.cpp]: we want our code to use `auto` always when possible
```suggestion
                const auto providerId =
                    model->data(model->index(i), OCC::UnifiedSearchResultsListModel::DataRole::ProviderIdRole)
                        .toString();

                const auto subline = 
                    model->data(model->index(i), OCC::UnifiedSearchResultsListModel::DataRole::SublineRole)
                        .toString();

                const auto title = 
                    model->data(model->index(i), OCC::UnifiedSearchResultsListModel::DataRole::TitleRole)
                        .toString();
```

File: src/gui/tray/unifiedsearchresultslistmodel.cpp [hunk 2]
```diff
@@ -324,26 +324,46 @@ bool UnifiedSearchResultsListModel::isSearchInProgress() const
     return !_searchJobConnections.isEmpty();
 }
 
-void UnifiedSearchResultsListModel::resultClicked(const QString &providerId, const QUrl &resourceUrl) const
+
+void UnifiedSearchResultsListModel::resultClicked(
+    const QString &providerId, const QUrl &resourceUrl, const QString &subline, const QString &title
+) const
```
Comment (mgallien) [inline src/gui/tray/unifiedsearchresultslistmodel.cpp]: I know our code style might not be well defined but arguments should not start on a new line (as visible for other methods)
```suggestion
void UnifiedSearchResultsListModel::resultClicked(const QString &providerId,
                                                                                           const QUrl &resourceUrl,
                                                                                           const QString &subline,
                                                                                           const QString &title) const
```

File: src/gui/tray/unifiedsearchresultslistmodel.cpp [hunk 3]
```diff
@@ -324,26 +324,46 @@ bool UnifiedSearchResultsListModel::isSearchInProgress() const
...
-    const auto dir = urlQuery.queryItemValue(QStringLiteral("dir"), QUrl::ComponentFormattingOption::FullyDecoded);
-    const auto fileName =
-        urlQuery.queryItemValue(QStringLiteral("scrollto"), QUrl::ComponentFormattingOption::FullyDecoded);
+    QString dir = urlQuery.queryItemValue(QStringLiteral("dir"), QUrl::ComponentFormattingOption::FullyDecoded);
+    QString fileName = urlQuery.queryItemValue(QStringLiteral("scrollto"), QUrl::ComponentFormattingOption::FullyDecoded);
 
-    if (providerId.contains(QStringLiteral("file"), Qt::CaseInsensitive) && !dir.isEmpty() && !fileName.isEmpty()) {
-        if (!_accountState || !_accountState->account()) {
+    if (providerId.contains(QStringLiteral("file"), Qt::CaseInsensitive)){
+        if (_accountState == nullptr || _accountState->account() == nullptr) {
```
Comment (mgallien) [inline src/gui/tray/unifiedsearchresultslistmodel.cpp]: please revert
the reader should not need to know that what you test are naked pointers
they could or maybe are smart pointers
it is good practice in c++ to test the validity of a pointer using the version of the code that you have modified (`if (!ptr) {`)
```suggestion
        if (!_accountState || !_accountState->account()) {
```

File: src/gui/tray/unifiedsearchresultslistmodel.h [hunk 4]
```diff
@@ -71,7 +71,7 @@ class UnifiedSearchResultsListModel : public QAbstractListModel
     [[nodiscard]] QString errorString() const;
     [[nodiscard]] bool waitingForSearchTermEditEnd() const;
 
-    Q_INVOKABLE void resultClicked(const QString &providerId, const QUrl &resourceUrl) const;
+    Q_INVOKABLE void resultClicked(const QString &providerId, const QUrl &resourceUrl, const QString &subline, const QString &title) const;
```
Comment (mgallien) [inline src/gui/tray/unifiedsearchresultslistmodel.h:74]:
```suggestion
    Q_INVOKABLE void resultClicked(const QString &providerId, const QUrl &resourceUrl, const QString &subLine, const QString &title) const;
```
we use camel case style

File: src/gui/tray/unifiedsearchresultslistmodel.cpp [hunk 5]
```diff
@@ -324,26 +324,46 @@ bool UnifiedSearchResultsListModel::isSearchInProgress() const
...
-        const auto localFiles =
-            FolderMan::instance()->findFileInLocalFolders(QFileInfo(relativePath).path(), _accountState->account());
+        QString relativePath;
+        // server version above 20
+        if (dir.isEmpty() && fileName.isEmpty()) {
+            // file is direct child of syncfolder
+            if (subline.isEmpty()) {
+                dir = title;
+            }
+            else {
```
Comment (mgallien) [inline src/gui/tray/unifiedsearchresultslistmodel.cpp]:
```suggestion
            } else {
```

File: src/gui/tray/unifiedsearchresultslistmodel.cpp [hunk 6]
```diff
@@ -324,26 +324,46 @@ bool UnifiedSearchResultsListModel::isSearchInProgress() const
...
+                dir = subline.split(' ', Qt::SkipEmptyParts).last();
+                fileName = QLatin1Char('/') + title;
+            }
+        }
+        // server version 20
+        else if (dir.length() > 1) {
```
Comment (mgallien) [inline src/gui/tray/unifiedsearchresultslistmodel.cpp]:
```suggestion
        } else if (dir.length() > 1) {
        // server version 20
```

File: src/gui/tray/unifiedsearchresultslistmodel.cpp [hunk 7]
```diff
@@ -324,26 +324,46 @@ bool UnifiedSearchResultsListModel::isSearchInProgress() const
...
+            fileName.prepend(QLatin1Char('/'));
+        }
+        relativePath = dir + fileName;
```
Comment (mgallien) [inline src/gui/tray/unifiedsearchresultslistmodel.cpp]:
```suggestion
        // server version above 20
        if (dir.isEmpty() && fileName.isEmpty()) {
            // file is direct child of syncfolder
            if (subline.isEmpty()) {
                dir = title;
            }
            else {
                dir = subline.split(' ', Qt::SkipEmptyParts).last();
                fileName = QLatin1Char('/') + title;
            }
        }
        // server version 20
        else if (dir.length() > 1) {
            fileName.prepend(QLatin1Char('/'));
        }
        const auto relativePath = dir + fileName;
```

File: src/gui/tray/unifiedsearchresultslistmodel.cpp [hunk 8]
```diff
@@ -324,26 +324,46 @@ bool UnifiedSearchResultsListModel::isSearchInProgress() const
...
 
+        const QStringList localFiles = FolderMan::instance()->findFileInLocalFolders(relativePath, _accountState->account());
```
Comment (mgallien) [inline src/gui/tray/unifiedsearchresultslistmodel.cpp]:
```suggestion
        const auto localFiles = FolderMan::instance()->findFileInLocalFolders(relativePath, _accountState->account());
```

File: src/gui/tray/unifiedsearchresultslistmodel.cpp [hunk 9]
```diff
@@ -324,26 +324,46 @@ bool UnifiedSearchResultsListModel::isSearchInProgress() const
...
         if (!localFiles.isEmpty()) {
-            qCInfo(lcUnifiedSearch) << "Opening file:" << localFiles.constFirst();
-            QDesktopServices::openUrl(QUrl::fromLocalFile(localFiles.constFirst()));
-            return;
+            qCInfo(lcUnifiedSearch) << "Opening file: " << localFiles.constFirst();
+            const bool fileOpenedLocally = QDesktopServices::openUrl(QUrl::fromLocalFile(localFiles.constFirst()));
+            if (fileOpenedLocally) {
+                return;
+            }
+            qCWarning(lcUnifiedSearch) << "Warning: QDesktopServices::openUrl unexpectedly failed to open the file. Opening resourceUrl in web browser is attempted next.";
```
Comment (mgallien) [inline src/gui/tray/unifiedsearchresultslistmodel.cpp]:
```suggestion
            if (fileOpenedLocally) {
                return;
            } else {
                qCWarning(lcUnifiedSearch) << "Warning: QDesktopServices::openUrl unexpectedly failed to open the file. Opening resourceUrl in web browser is attempted next.";
            }
```

File: src/gui/tray/unifiedsearchresultslistmodel.cpp [hunk 10]
```diff
@@ -324,26 +324,46 @@ bool UnifiedSearchResultsListModel::isSearchInProgress() const
...
+void UnifiedSearchResultsListModel::resultClicked(
+    const QString &providerId, const QUrl &resourceUrl, const QString &subline, const QString &title
+) const
 {
     const QUrlQuery urlQuery{resourceUrl};
-    const auto dir = urlQuery.queryItemValue(QStringLiteral("dir"), QUrl::ComponentFormattingOption::FullyDecoded);
-    const auto fileName =
-        urlQuery.queryItemValue(QStringLiteral("scrollto"), QUrl::ComponentFormattingOption::FullyDecoded);
+    QString dir = urlQuery.queryItemValue(QStringLiteral("dir"), QUrl::ComponentFormattingOption::FullyDecoded);
+    QString fileName = urlQuery.queryItemValue(QStringLiteral("scrollto"), QUrl::ComponentFormattingOption::FullyDecoded);
```
Comment (mgallien) [inline src/gui/tray/unifiedsearchresultslistmodel.cpp]:
```suggestion
    const auto dir = urlQuery.queryItemValue(QStringLiteral("dir"), QUrl::ComponentFormattingOption::FullyDecoded);
    const auto fileName = urlQuery.queryItemValue(QStringLiteral("scrollto"), QUrl::ComponentFormattingOption::FullyDecoded);
```

File: src/gui/tray/unifiedsearchresultslistmodel.cpp [hunk 11]
```diff
@@ -324,26 +324,46 @@ bool UnifiedSearchResultsListModel::isSearchInProgress() const
...
+        }
+        relativePath = dir + fileName;
 
+        const QStringList localFiles = FolderMan::instance()->findFileInLocalFolders(relativePath, _accountState->account());
         if (!localFiles.isEmpty()) {
-            qCInfo(lcUnifiedSearch) << "Opening file:" << localFiles.constFirst();
-            QDesktopServices::openUrl(QUrl::fromLocalFile(localFiles.constFirst()));
-            return;
+            qCInfo(lcUnifiedSearch) << "Opening file: " << localFiles.constFirst();
+            const bool fileOpenedLocally = QDesktopServices::openUrl(QUrl::fromLocalFile(localFiles.constFirst()));
```
Comment (mgallien) [inline src/gui/tray/unifiedsearchresultslistmodel.cpp]:
```suggestion
            const auto fileOpenedLocally = QDesktopServices::openUrl(QUrl::fromLocalFile(localFiles.constFirst()));
```

**Review by mgallien — CHANGES_REQUESTED** [review]
see my inline comments
can you also make sure that there are new automated tests covering the new code branches ?
we will need to have tests to understand if the changes are working as expected

File: test/testunifiedsearchlistmodel.cpp (hunk 1 above)
Comment (barisbasar1209) [reply test/testunifiedsearchlistmodel.cpp]: Noted. Thank you! I actually noticed that auto was used quite often and I thought this was rather unclean. 
Could you explain to me why `auto` is preferred?

Comment (mgallien) [reply test/testunifiedsearchlistmodel.cpp]: the main reason to use `auto`is the DRY concept
Do not Repeat Yourself
this makes also the maintenance of the code much easier and often makes the code more readable
for reference and completeness see https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/

File: src/gui/tray/unifiedsearchresultslistmodel.cpp (hunk 3 above)
Comment (barisbasar1209) [reply src/gui/tray/unifiedsearchresultslistmodel.cpp]: Thank you, didn't know this.

File: src/gui/tray/unifiedsearchresultslistmodel.h (hunk 4 above)
Comment (barisbasar1209) [reply src/gui/tray/unifiedsearchresultslistmodel.h:74]: The attribute was called `subline` everywhere else I found it. Also in the server documentation it is called `subline`. 
Should it still be adusted?
image

File: src/gui/tray/unifiedsearchresultslistmodel.cpp (hunk 7 above)
Comment (barisbasar1209) [reply src/gui/tray/unifiedsearchresultslistmodel.cpp]: Ah yes, thanks for spotting this :)

File: src/gui/tray/unifiedsearchresultslistmodel.cpp (hunk 10 above)
Comment (barisbasar1209) [reply src/gui/tray/unifiedsearchresultslistmodel.cpp]: I understand the `auto` part but wouldn't the `const` collide with the way I implemented the logic, since there  I assign to `dir` and `fileName` ?

File: src/gui/tray/unifiedsearchresultslistmodel.h (hunk 4 above)
Comment (mgallien) [reply src/gui/tray/unifiedsearchresultslistmodel.h:74]: no it is fine
thanks for the API doc, I would not have checked this

File: src/gui/tray/unifiedsearchresultslistmodel.cpp (hunk 10 above)
Comment (mgallien) [reply src/gui/tray/unifiedsearchresultslistmodel.cpp]: I did not checked if those variables could be constant or not
I just wanted to point out that by default they should be defined like this: `const auto variableName = value;`

**Comment by barisbasar1209** [comment]
@mgallien I have two more questions: 

1. I used the `sign off and commit suggestions` button to accept your suggestions. Is this the preferred way to go? (Maybe I should have asked in advance, I am sorry)
2. Are the parameters in [64f2b0b](https://github.com/nextcloud/desktop/pull/8575/commits/64f2b0b8390b2e75249ffe8e1f05f652664d0ba9) really supposed to be formatted this way?

Thanks for your help!

File: src/gui/tray/unifiedsearchresultslistmodel.cpp (hunk 7 above)
Comment (barisbasar1209) [reply src/gui/tray/unifiedsearchresultslistmodel.cpp]: after I signed off and committed the other suggestions this one became outdated so I manually commit the suggested changes, I hope that's fine

**Comment by mike0609king** [comment]
@mgallien The automated tests have been implemented; the newly implemented code has 100% coverage. We have quite a few edge cases in terms of what nextcloud returns in the `subline` and `title` section. Furthermore, the behavior is different for nextcloud version 20 and nextcloud version 20+.

We also decided to make tests independent of the results that have been defined in `initSearchResultData`, because 
1. we wanted to avoid breaking already implemented tests,
2. the existing data in the tests is very redundant for our test cases; same entries in `subline` and `title` and
3. we wanted to avoid to do the typical setup of those tests, i.e. `setSearchTerm` and `wait. This would make the code less readable.

**Comment by mike0609king** [comment]
Forgot to restore old accountState value after setting it to a null pointer. This caused tests after `testSearchResultClicked` to fail. IS fixed now.

**Comment by camilasan** [comment]
We see so many potential issues with virtual files on demand, so we decided to close the PR.
thanks for the efforts on it though.

**Comment by mijioij** [comment]
@camilasan Please clarify what this pr does with virtual files and what primary issues you foresee. This PR was meant to allow searched files using tray application to be opened using local explorer when all versions of nextcloud server are used and not just old server versions .
